### ISSUE TITLE
[Feature] Add Prestige Mode

### ISSUE THREAD

**Issue by francktrouillez** [pagefaultgames/pokerogue#1545]
## What are Prestiges?

The idea of the **Prestiges** is to allow the player to start over the _Classic_ mode with some increased difficulties. The game will include some modifiers to make the game harder. It contains multiple levels of difficulty, and each level will be harder than the previous one. You need to complete the game in the current difficulty to unlock the next one.

To unlock **Prestiges**, you first need to complete the game in the _Classic_ mode. Once you've done that, you will be able to start over the game in the _Prestige_ mode.

## Why introducing Prestiges?

I found a [very interesting ticket talking](https://github.com/pagefaultgames/pokerogue/issues/1283) about this kind of system, which I believe makes a lot of sense in roguelite games. We can find similar systems in games like [_Slay The Spire_](https://slay-the-spire.fandom.com/wiki/Ascension) for instance.

Also, I really like the idea of adding difficulty and challenges to the game, without going for the endless mode. I believe this can bring a lot of replayability to the game, and I'm sure some players will enjoy this kind of challenge.

## Why would someone play Prestiges?

Short answer: for the PRESTIGE!

On another note, I believe it makes it quite exciting to start over the game with some increased difficulties.
I personally prefer games with challenges, than games that are using the length of the game as a way to make it harder. Of course, this is personal, but I believe some players will enjoy this kind of challenge.

It can be a good way to challenge yourself and see how far you can go in the game. What is nice about this system is that we can potentially extend it with more levels, or mode modifiers if needed, which means more difficulties/challenges.

## What is the plan?

If this idea makes sense, I'm proposing the following plan to introduce the Prestiges.

I'd like to make use of a feature flag to enable/disable the Prestiges, so we can incrementally work on it out before releasing it to the public. Indeed, I believe this is quite a broad project, and I'm not comfortable releasing it in one go. I also believe it will be much easier for people to review the changes if they are split into smaller PRs, and to actually contribute to the project.

Here is the plan I'm proposing:
- [ ] Add a feature flag to enable/disable the Prestiges, currently set to `false` by default: https://github.com/pagefaultgames/pokerogue/pull/1533
- [ ] **Add foundations for the Prestiges system**: https://github.com/pagefaultgames/pokerogue/pull/1534
  - Add UI improvements
    - [ ] Add a Prestige option select whenever the player chooses `Classic` mode: https://github.com/pagefaultgames/pokerogue/pull/1540
    - [ ] Add the current Prestige level in the UI when playing the game: https://github.com/pagefaultgames/pokerogue/pull/1541
    - [ ] Add the maximum Prestige level completed in the Game stats: https://github.com/pagefaultgames/pokerogue/pull/1542
    - [ ] Add the Prestige level in the save slot: https://github.com/pagefaultgames/pokerogue/pull/1543
    - [ ] Add Prestige achievements for each level completed: TODO
    - [ ] Add locales for the Prestige mode: TODO
    - ...
  - Add Prestige modifiers
    - [ ] Decrease the starter party points: https://github.com/pagefaultgames/pokerogue/pull/1535
    - [ ] Increase enemy pokemons stats: https://github.com/pagefaultgames/pokerogue/pull/1536
    - [ ] Increase the shop item prices: https://github.com/pagefaultgames/pokerogue/pull/1537
    - [ ] Decrease the party luck: https://github.com/pagefaultgames/pokerogue/pull/1538
    - [ ] Decrease the experience gain: https://github.com/pagefaultgames/pokerogue/pull/1539
    - ...
- [ ] Enable the Prestiges feature flag, by setting it to `true`: https://github.com/pagefaultgames/pokerogue/pull/1544
- [ ] Remove the feature flag: TODO

The idea of this plan is to make each PR really small, easy-to-review, and as self-contained as possible. I believe this will help to have a better understanding of the changes.

Also, by doing so, anyone can actually start contributing to the project by starting a branch from the **_Add foundations for the Prestiges system_** PR, whether it is for adding new modifiers, achievements, UI improvements, or anything else.

By going this way, it means we don't need to have everything I listed to consider the system as ready. We can discard any PR if we believe it doesn't make sense, or if we believe it can be improved, or add any new ones if needed.

We can consider the Prestiges system as ready whenever we want to actually. We can incrementally add new features to it under the feature flag, but we can also decide to release it once it has some minimal features.

Of course, this plan is just a suggestion, and I'm open to any feedback on it. I believe this can be a good way to introduce the Prestiges system to the game.

Once again, I strongly believe this can be a good addition to the game if that makes sense.

## What could it look like?

I've been working on some of the UI improvements and modifiers for the Prestiges system. Considering the PRs I've been working on, here is what it could look like:
- Unlocking the Prestige mode

  https://github.com/pagefaultgames/pokerogue/assets/57403591/2e441484-c471-4dac-98d6-1cb14d610f5b

- Unlocking the next Prestige level

  https://github.com/pagefaultgames/pokerogue/assets/57403591/e879e2e0-b55a-4646-a06e-c3b3e94e4202

- Starting a new game with the Prestige level selection
  (Starters party points decreased, enemy pokemons stats increased, shop item prices increased, party luck decreased, experience gain decreased)

  https://github.com/pagefaultgames/pokerogue/assets/57403591/cf984eca-2d87-4247-8689-f1bb6e6b1804

Note: I didn't dive into level design, so we might want to tweak the values of the modifiers.

## Thoughts?

What do you guys think?

### Edit
I just saw that @Xavion3  was also working on a similar system https://github.com/pagefaultgames/pokerogue/pull/1459
I decided to still move forward with unveiling what I was working on these last days, as I believe we might want to merge some ideas together, since I have the feeling they are quite similar. I'm open to any feedback on this. Of course, @Xavion3, you are part of the main developers, so do whatever you think is best for the game. Feel free to just modify/close any PRs you believe are not relevant. Or to just discard this project.

**I really hope this can help!**

**Comment by Xavion** [comment]
Sorry, you have emailed the wrong Xavion.

**Comment by molker** [comment]
Hello! Love to see the passion in wanting to add such a big feature like this. I would appreciate if you could combine your PRs into one. 
We already have 150+ other PRs we need to look at, adding 13 PRs for one feature only adds to the flood.

**Comment by Xavion** [comment]
Sorry, you have (once again) emailed the wrong Xavion.  Can you please take
me off the list somehow?  As you're emailing me directly, it won't let me
unsubscribe.

**Comment by molker** [comment]
> Sorry, you have (once again) emailed the wrong Xavion.  Can you please take
me off the list somehow?  As you're emailing me directly, it won't let me
unsubscribe.

you have to open the website and unsubscribe from the notifications on the right side of the screen 

>  Do you think all the features described make sense? I can pop off some of them, just to avoid requiring someone to read a code that doesn't bring any value.

I haven't had enough time to look through them, I can try and get to that tomorrow. I primarily saw all the PRs pop up.

**Comment by Xavion** [comment]
>
> you have to open the website and unsubscribe from the notifications on the
> right side of the screen
>
In my last reply, I wrote the following: "*As you're emailing me directly*,
it won't let me unsubscribe."  This means I already attempted to
unsubscribe (using the method you have mentioned).  It won't let me do so
as you're emailing me directly (i.e. to my Gmail address).  Take a look at
the headers of the emails to see what I mean.  The only way I can be
unsubscribed is for one of you to take me off the list manually.  The
GItHub website (erroneously) thinks I am already unsubscribed.  If I click
the button on the right side of the screen (as you said), it will subscribe
me twice!
